trinity/sync/full/chain.py

```diff
@@ -442,29 +380,444 @@ class FastChainSyncer(BaseHeaderChainSyncer):
+            ))
+            self.header_queue.complete(batch_id, headers)
+
+    async def _display_stats(self) -> None:
+        last_head = await self.wait(self.db.coro_get_canonical_head())
+        timer = Timer()
+
+        while self.is_operational:
+            await self.sleep(5)
+            self.logger.debug(
+                "(in progress, queued, max size) of headers, bodies, receipts: %r",
+                [(q.num_in_progress(), len(q), q._maxsize) for q in (
+                    self.header_queue,
+                    self._block_body_tasks,
+                    self._receipt_tasks,
+                )],
+            )
+
+            head = await self.wait(self.db.coro_get_canonical_head())
+            if head == last_head:
+                continue
+            else:
+                block_num_change = head.block_number - last_head.block_number
+                last_head = head
+
+                self.logger.info(
+                    "Advanced by %d blocks in %0.1f seconds, new head: #%d",
+                    block_num_change, timer.pop_elapsed(), head.block_number)
+
+    async def _persist_ready_blocks(self) -> None:
+        """
+        Persist blocks as soon as all their prerequisites are done: body and receipt downloads.
+        Persisting must happen in order, so that the block's parent has already been persisted.
+
+        Also, determine if fast sync with this peer should end, having reached (or surpassed)
+        its target hash. If so, shut down this service.
+        """
+        while self.is_operational:
+            # jhis tracker waits for all prerequisites to be complete, and returns headers in
+            # order, so that each header's parent is already persisted.
+            completed_headers = await self.wait(self._block_persist_tracker.ready_tasks())
+
+            await self._persist_blocks(completed_headers)
+
+            target_hash = self.get_target_header_hash()
+
+            if target_hash in [header.hash for header in completed_headers]:
+                # simply exit the service when reaching the target hash
+                self.cancel_nowait()
+                break
+
+    async def _persist_blocks(self, headers: Tuple[BlockHeader, ...]) -> None:
+        """
+        Persist blocks for the given headers, directly to the database
+
+        :param headers: headers for which block bodies and receipts have been downloaded
+        """
+        for header in headers:
+            if header.uncles_hash != EMPTY_UNCLE_HASH:
+                body = self._pending_bodies.pop(header)
+                uncles = body.uncles
+            else:
+                uncles = tuple()
+
+            vm_class = self.chain.get_vm_class(header)
+            block_class = vm_class.get_block_class()
+            # We don't need to use our block transactions here because persist_block() doesn't do
+            # anything with them as it expects them to have been persisted already.
```

Comment:
I don't think this comment is accurate:

https://github.com/ethereum/py-evm/blob/de4809fe77f9e7502d51478a30f443f007fd0923/eth/db/chain.py#L233-L243

By not including the transactions, we are failing to populate the canonical transaction hash to block lookups right?

Reply:
Hm, I haven't looked deeply (it was already there), but at first glance I think you're right.